README.md
## Introduction

A wrapper application to use gnome-bluetooth outside of GNOME.

Depends on gnome-bluetooth >= 3.14 and <= 41.

---

## How to build and install

#### Download the source code and enter the source directory

```
# Clone this repo:
git clone https://github.com/linuxmint/blueberry.git

# Enter the folder:
cd blueberry
```

#### Building the package

```
# Use mint-build to build the project for the first time.
# This also fetches and installs the build dependencies:
mint-build

# In subsequent builds, you can use dpkg-buildpackage for faster builds:
dpkg-buildPackage
```

#### Install the package:

```
# Once that succeeds, install:
cd ..
sudo dpkg -i blueberry\*.deb
```

For more information, refer the developer guide:
https://linuxmint-developer-guide.readthedocs.io/en/latest/index.html


debian/changelog
blueberry (1.4.8) vanessa; urgency=medium

  * Update README.md
  * l10n: Update files
  * l10n: Update translations

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 24 Jul 2022 11:52:22 +0200

blueberry (1.4.7) una; urgency=medium

  * l10n: Update translations

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 31 Dec 2021 12:33:32 +0000

blueberry (1.4.6) una; urgency=medium

  * l10n: Update translations
  * l10n: Update files

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 06 Dec 2021 15:22:10 +0000

blueberry (1.4.5) una; urgency=medium

  [ P. Radha Krishna ]
  * Added Turn bluetooth on/off option in tray menu #112 (#117)

  [ Michael Webster ]
  * blueberry-tray.py: Remove menu positioning code.
  * blueberry-tray.py: Silence a runtime warning.

  [ Mickey Rose ]
  * rfkillMagic: rewrite safechild in shell script to reduce memory usage (#115)

  [ P. Radha Krishna ]
  * Added build & install steps to readme file (#118)

  [ Clement Lefebvre ]
  * l10n: Update POT

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 26 Nov 2021 10:14:48 +0000

blueberry (1.4.4) uma; urgency=medium

  * l10n: Update translations
  * l10n: Generate files

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 25 Jun 2021 14:38:00 +0100

blueberry (1.4.3) uma; urgency=medium

  * l10n: Update translations
  * l10n: Generate files

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 10 Jun 2021 11:46:32 +0100

blueberry (1.4.2) ulyssa; urgency=medium

  * l10n: Update translations
  * l10n: Update files

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 02 Jan 2021 16:31:59 +0000

blueberry (1.4.1) ulyssa; urgency=medium

  * l10n: Update translations
  * l10n: Generate files

 -- Clement Lefebvre <root@linuxmint.com>  Thu, 10 Dec 2020 12:40:56 +0000

blueberry (1.4.0) ulyssa; urgency=medium

  [ okaestne ]
  * Don't use /usr/bin/env to call python3

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 30 Nov 2020 11:39:47 +0000

blueberry (1.3.9) ulyana; urgency=medium

  * l10n: Update translations

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 24 Jun 2020 11:39:08 +0100

blueberry (1.3.8) ulyana; urgency=medium

  [ Michael Webster ]
  * blueberry.ui: Change the valign of the headerbar switch to center, otherwise it wants to expand to fill the entire headerbar height.

  [ Clement Lefebvre ]
  * Remove Cinnamon applet
  * Make active tray icon more recognizable

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 25 May 2020 17:48:43 +0100

blueberry (1.3.7) ulyana; urgency=medium

  [ cptpcrd ]
  * obex: Remove hard requirement on deprecated hcitool

  [ Clement Lefebvre ]
  * l10n: Update translations
  * l10n: Generate desktop files

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 12 May 2020 10:08:32 +0100

blueberry (1.3.6) ulyana; urgency=medium

  [ Clement Lefebvre ]
  * l10n: Update POT

  [ Michael Webster ]
  * blueberry.ui: Lower version requirement (current releases are still on gtk 3.22.
  * Fix debian/control to refer to python3.

  [ Clement Lefebvre ]
  * tray: Switch to symbolic icons (#100)

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 21 Mar 2020 10:08:52 +0000

blueberry (1.3.5) debbie; urgency=medium

  [ Michael Webster ]
  * blueberry.py: Fix handling of the device name/alias.

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 04 Feb 2020 15:10:44 +0000

blueberry (1.3.4) tricia; urgency=medium

  * l10n: Update translations
  * l10n: Update files

 -- Clement Lefebvre <root@linuxmint.com>  Mon, 06 Jan 2020 11:10:29 +0000

blueberry (1.3.3) tricia; urgency=medium

  * l10n: Update translations

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 11 Dec 2019 13:55:49 +0000

blueberry (1.3.2) tricia; urgency=medium

  [ samtygier ]
  * Handle error from rfkill (#94)

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 10 Dec 2019 10:12:12 +0000

blueberry (1.3.1) tricia; urgency=medium

  * l10n: Update translations
  * l10n: Generate files

 -- Clement Lefebvre <root@linuxmint.com>  Tue, 26 Nov 2019 17:03:29 +0000

blueberry (1.3.0) tricia; urgency=medium

  [ Clement Lefebvre ]
  * Add support for XAppStatusIcon
  * StatusIcon: Switch to activate signal

  [ James Ross ]
  * obex: check for None when setting shared path
  * rework override to be compatible with any version

  [ shanmukhateja ]
  * Fix tray icon not shown with window (#89)

  [ James Ross ]
  * blueberry: UI redesign
  * rfkillMagic: support hotpluggable bt devices
  * rfkillMagic: support device indexes above 9
  * blueberry: rework rfkill error to be switch-specific
  * cleanup some unneded code

  [ Clement Lefebvre ]
  * UI: Ensure burger menu is visible
  * l10n: Generate files
  * l10n: Update POT

 -- Clement Lefebvre <root@linuxmint.com>  Fri, 22 Nov 2019 14:16:18 +0000

blueberry (1.2.9) tina; urgency=medium

  * l10n: Update translations

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 28 Jul 2019 15:23:07 +0200

blueberry (1.2.8) tina; urgency=medium

  [ Riccardo ]
  * Don't spin spinner in devices tab (#85)

 -- Clement Lefebvre <root@linuxmint.com>  Sat, 20 Jul 2019 16:48:45 +0200

blueberry (1.2.7) tina; urgency=medium

  * l10n: Update translations

 -- Clement Lefebvre <root@linuxmint.com>  Wed, 10 Jul 2019 11:03:57 +0200

blueberry (1.2.6) tina; urgency=medium

  [ Clement Lefebvre ]
  * Fix generate_desktop_files

  [ Sebastiaan Lokhorst ]
  * Convert everything to use Python 3 (#66)

  [ Koen ]
  * Add paired submenu for quick connect/disconnect (#69)

  [ Clement Lefebvre ]
  * Tray: Use a tick image to represent connected device
  * l10n: Update POT
  * Tray: Fix paired menu when no devices are paired
  * Use shell launchers and use setproctitle to name the python processes

  [ shanmukhateja ]
  * rfkill: Show error message on permission denied (#83)

  [ Clement Lefebvre ]
  * rfkill: Show the error message in the UI
  * l10n: Update POT

 -- Clement Lefebvre <root@linuxmint.com>  Sun, 30 Jun 2019 17:55:33 +0200

blueberry (1.2.5) tessa; urgency=medium

  * l10n: Update translations


usr/lib/blueberry/btadapter.py

import subprocess
import re


class BtAdapter:
    def get_discoverable(self):
        try:
            output = subprocess.check_output(["bt-adapter", "-i"], timeout=2).decode("utf-8").strip()
            match = re.search("Discoverable: ([0|1]) \[rw\]", output)
            if match:
                return match.group(1) == '1'
        except Exception as e:
            pass
        return False

    def get_powered(self):
        try:
            output = subprocess.check_output(["bt-adapter", "-i"], timeout=2).decode("utf-8").strip()
            match = re.search("Powered: ([0|1]) \[rw\]", output)
            if match:
                return match.group(1) == '1'
        except Exception as e:
            pass
        return False

    def set_discoverabletimeout(self):
        try:
            output = subprocess.run(["bt-adapter", "--set", "DiscoverableTimeout", "60"], timeout=2, check=True,
                                    capture_output=True)
        except Exception:
            pass

    def poweron(self):
        try:
            output = subprocess.run(["bt-adapter", "--set", "Powered", "1" ], timeout=2, check=True,
                                    capture_output=True)
        except Exception:
            pass

    def make_discoverable(self, visible=True):
        try:
            output = subprocess.run(["bt-adapter", "--set", "Discoverable", "1" if visible else "0"], timeout=2, check=True,
                                    capture_output=True)
        except Exception:
            pass


usr/lib/blueberry/rfkillMagic.py

import _thread as thread
import subprocess
import re
from gi.repository import GLib

RFKILL_CHK = ["/usr/sbin/rfkill", "list", "bluetooth"]
RFKILL_BLOCK = ["/usr/sbin/rfkill", "block", "bluetooth"]
RFKILL_UNBLOCK = ["/usr/sbin/rfkill", "unblock", "bluetooth"]

RFKILL_EVENT_MONITOR = ["/usr/lib/blueberry/safechild", "/usr/sbin/rfkill", "event"]

class Interface:
    def __init__(self, output_callback, debug):
        self.enable_debugging = debug
        self.output_callback = output_callback
        self.have_adapter = False
        self.adapter_index = -1

        self.tproc = None
        self.blockproc = None

        self.hard_block = False
        self.soft_block = False
        self.rfkill_err = None

        self.monitor_killer = False

        self.adapter_check()
        self.start_event_monitor()

    def adapter_check(self):
        proc = subprocess.run(RFKILL_CHK, stdout=subprocess.PIPE)
        if proc.returncode != 0:
            self.debug("Error running command: %s." % RFKILL_CHK)
            res = ""
        else:
            res = proc.stdout.decode('utf-8')

        match = None
        have_adapter = False

        '''
        Assume the output of:

        > /usr/sbin/rfkill list bluetooth

        looks like:

        1: hci0: Bluetooth
            Soft blocked: yes
            Hard blocked: no
        '''
        if res:
            match = re.search(r'^(?P<idx>\d+): .+: Bluetooth\n', res)

        if match:
            self.debug("adapter_check full output:\n%s" % res)
            self.adapter_index = int(match.group('idx'))
            self.debug("adapter_check found adapter at %d" % self.adapter_index)
            have_adapter = True
        else:
            self.debug("adapter_check no output (no adapter)")

        self.have_adapter = have_adapter
        return have_adapter

    def start_event_monitor(self):
        if not self.tproc and not self.monitor_killer:
            thread.start_new_thread(self.event_monitor_thread, (None,))

    def event_monitor_thread(self, data):
        self.tproc = subprocess.Popen(RFKILL_EVENT_MONITOR, stdout=subprocess.PIPE, stdin=subprocess.PIPE)
        while self.tproc.poll() is None and not self.monitor_killer:
            l = self.tproc.stdout.readline().decode('utf-8') # This blocks until it receives a newline.
            self.update_state(l)

    def update_state(self, line):
        self.debug("update_state line: %s" % line)

        '''
        Assume the output of:

        > /usr/bin/rfkill event

        looks like:

        1426095957.906704: idx 0 type 1 op 0 soft 0 hard 0
        1426095957.906769: idx 1 type 2 op 0 soft 1 hard 0
        1426096013.465033: idx 1 type 2 op 2 soft 0 hard 0

        or:

        2017-12-08 11:54:16,972291-0800: idx 0 type 2 op 0 soft 0 hard 0
        2017-12-08 11:54:16,972431-0800: idx 1 type 1 op 0 soft 0 hard 0
        2017-12-08 11:54:16,972474-0800: idx 4 type 2 op 0 soft 0 hard 0
        '''
        if not self.have_adapter:
            self.adapter_check()

        if self.have_adapter:
            match = re.search(r'idx (?P<idx>\d+) type (?P<type>\d+) op (?P<op>\d+) soft (?P<soft>\d+) hard (?P<hard>\d+)', line)
            if match:
                if int(match.group('idx')) == self.adapter_index:
                    if int(match.group('op')) == 1:
                        self.adapter_check()
                    self.soft_block = int(match.group('soft')) == 1
                    self.hard_block = int(match.group('hard')) == 1

        self.update_ui()

    def update_ui(self):
        GLib.idle_add(self.output_callback)

    def try_set_blocked(self, blocked):
        thread.start_new_thread(self.set_block_thread, (blocked,))

    def set_block_thread(self, data):
        block = data

        if block:
            self.debug("set_block_thread blocking")
            self.blockproc = subprocess.Popen(RFKILL_BLOCK, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        else:
            self.debug("set_block_thread unblocking")
            self.blockproc = subprocess.Popen(RFKILL_UNBLOCK, stdout=subprocess.PIPE, stderr=subprocess.PIPE)

        # Check for errors and continue
        _,err = self.blockproc.communicate()
        if err:
            error = err.decode("utf-8")
            self.debug(error)
            self.rfkill_err = error
            # Force UI update
            self.update_ui()
        else:
            self.rfkill_err = None

        self.blockproc = None
        self.debug("set_block_thread finished")
        thread.exit()

    def terminate(self):
        if self.blockproc:
            self.blockproc.kill()

    def debug(self, msg):
        if self.enable_debugging:
            print(msg)



usr/lib/blueberry/blueberry-tray.py
#!/usr/bin/python3

import sys
import gettext
import gi
gi.require_version('Gtk', '3.0')
gi.require_version('GnomeBluetooth', '1.0')
gi.require_version('XApp', '1.0')
from gi.repository import Gtk, Gdk, GnomeBluetooth, Gio, XApp
import rfkillMagic
import setproctitle
import subprocess

# i18n
gettext.install("blueberry", "/usr/share/locale")

setproctitle.setproctitle("blueberry-tray")

class BluetoothTray(Gtk.Application):
    def __init__(self):
        super(BluetoothTray, self).__init__(register_session=True, application_id="org.linuxmint.blueberry.tray")

    def do_activate(self):
        self.hold()

    def do_startup(self):
        Gtk.Application.do_startup(self)

        debug = False
        if len(sys.argv) > 1 and sys.argv[1] == "debug":
            debug = True

        self.rfkill = rfkillMagic.Interface(self.update_icon_callback, debug)
        self.settings = Gio.Settings(schema="org.blueberry")
        self.settings.connect("changed::tray-enabled", self.on_settings_changed_cb)

        self.tray_icon = "blueberry-tray"
        self.tray_active_icon = "blueberry-tray-active"
        self.tray_disabled_icon = "blueberry-tray-disabled"
        if self.settings.get_boolean("use-symbolic-icons"):
            self.tray_icon = "blueberry-tray-symbolic"
            self.tray_active_icon = "blueberry-tray-active-symbolic"
            self.tray_disabled_icon = "blueberry-tray-disabled-symbolic"

        # If we have no adapter, or disabled tray, end early
        if (not self.rfkill.have_adapter) or (not self.settings.get_boolean("tray-enabled")):
            self.rfkill.terminate()
            sys.exit(0)

        self.client = GnomeBluetooth.Client()
        self.model = self.client.get_model()
        self.model.connect('row-changed', self.update_icon_callback)
        self.model.connect('row-deleted', self.update_icon_callback)
        self.model.connect('row-inserted', self.update_icon_callback)

        self.icon = XApp.StatusIcon()
        self.icon.set_name("blueberry")
        self.icon.set_tooltip_text(_("Bluetooth"))
        self.icon.connect("activate", self.on_statusicon_activated)
        self.icon.connect("button-release-event", self.on_statusicon_released)

        self.update_icon_callback(None, None, None)

    def on_settings_changed_cb(self, setting, key, data=None):
        if not self.settings.get_boolean("tray-enabled"):
            self.terminate()

    def update_icon_callback(self, path=None, iter=None, data=None):
        if not self.rfkill.have_adapter:
            self.terminate(None)
            return

        if self.rfkill.hard_block or self.rfkill.soft_block:
            self.icon.set_icon_name(self.tray_disabled_icon)
            self.icon.set_tooltip_text(_("Bluetooth is disabled"))
        else:
            self.icon.set_icon_name(self.tray_icon)
            self.update_connected_state()

    def update_connected_state(self):
        self.get_devices()

        if len(self.connected_devices) > 0:
            self.icon.set_icon_name(self.tray_active_icon)
            self.icon.set_tooltip_text(_("Bluetooth: Connected to %s") % (", ".join(self.connected_devices)))
        else:
            self.icon.set_icon_name(self.tray_icon)
            self.icon.set_tooltip_text(_("Bluetooth"))

    def get_devices(self):
        self.connected_devices = []
        self.paired_devices = {}
        default_iter = None

        iter = self.model.get_iter_first()

        while iter:
            default = self.model.get_value(iter, GnomeBluetooth.Column.DEFAULT)
            if default:
                default_iter = iter
                break
            iter = self.model.iter_next(iter)

        if default_iter != None:
            iter = self.model.iter_children(default_iter)
            while iter:
                connected = self.model.get_value(iter, GnomeBluetooth.Column.CONNECTED)
                if connected:
                    name = self.model.get_value(iter, GnomeBluetooth.Column.NAME)
                    self.connected_devices.append(name)

                paired = self.model.get_value(iter, GnomeBluetooth.Column.PAIRED)
                if paired:
                    name = self.model.get_value(iter, GnomeBluetooth.Column.NAME)
                    proxy = self.model.get_value(iter, GnomeBluetooth.Column.PROXY)
                    self.paired_devices[name] = proxy

                iter = self.model.iter_next(iter)

    def on_statusicon_activated(self, icon, button, time):
        if button == Gdk.BUTTON_PRIMARY:
            subprocess.Popen(["blueberry"])

    def on_statusicon_released(self, icon, x, y, button, time, position):
        if button == 3:
            menu = Gtk.Menu()

            if not self.rfkill.hard_block:
                if self.rfkill.soft_block:
                    item = Gtk.MenuItem(label=_("Turn on Bluetooth"))
                    item.connect("activate", self.turn_on_bluetooth)
                    menu.append(item)
                else:
                    item = Gtk.MenuItem(label=_("Turn off Bluetooth"))
                    item.connect("activate", self.turn_off_bluetooth)
                    menu.append(item)

            if not(self.rfkill.hard_block or self.rfkill.soft_block):
                item = Gtk.MenuItem(label=_("Send files to a device"))
                item.connect("activate", self.send_files_cb)
                menu.append(item)

            item = Gtk.MenuItem(label=_("Open Bluetooth device manager"))
            item.connect("activate", self.open_manager_cb)
            menu.append(item)

            if len(self.paired_devices) > 0:
                menu.append(Gtk.SeparatorMenuItem())
                m_item = Gtk.MenuItem(label=_("Paired devices"))
                menu.append(m_item)
                paired_menu = Gtk.Menu()
                m_item.set_submenu(paired_menu)
                for device in self.paired_devices:
                    label = device
                    item = Gtk.ImageMenuItem(label=label)
                    if device in self.connected_devices:
                        image = Gtk.Image.new_from_icon_name("emblem-ok-symbolic", Gtk.IconSize.MENU)
                        image.set_tooltip_text(_("Connected"))
                        item.set_always_show_image(True)
                        item.set_image(image)
                    item.connect("activate",self.toggle_connect_cb, device)
                    paired_menu.append(item)

            menu.append(Gtk.SeparatorMenuItem())

            item = Gtk.MenuItem(label=_("Quit"))
            item.connect("activate", self.terminate)
            menu.append(item)

            menu.show_all()
            icon.popup_menu(menu, x, y, button, time, position)

    def toggle_connect_cb(self, item, data = None):
        proxy = self.paired_devices[data]
        connected = data in self.connected_devices
        self.client.connect_service(proxy.get_object_path(), not connected)

    def send_files_cb(self, item, data = None):
        subprocess.Popen(["bluetooth-sendto"])

    def open_manager_cb(self, item, data = None):
        subprocess.Popen(["blueberry"])

    def turn_on_bluetooth(self, item):
        self.rfkill.try_set_blocked(False)
        return True

    def turn_off_bluetooth(self,item):
        self.rfkill.try_set_blocked(True)
        return True

    def terminate(self, window = None, data = None):
        self.quit()

if __name__ == "__main__":
    BluetoothTray().run()


usr/lib/blueberry/blueberry.py
#!/usr/bin/python3

import sys, os
import gettext
import rfkillMagic
import setproctitle
import subprocess
from BlueberrySettingsWidgets import SettingsBox, SettingsRow
import gi
gi.require_version('Gtk', '3.0')
gi.require_version('GnomeBluetooth', '1.0')
from gi.repository import Gtk, GnomeBluetooth, Gio, GLib

SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))
APPLICATION_ID = 'com.linuxmint.blueberry'

# i18n
gettext.install("blueberry", "/usr/share/locale")

setproctitle.setproctitle("blueberry")

# similar to g_warning by default, can specify any log level though
def log(text, log_level=GLib.LogLevelFlags.LEVEL_WARNING):
    variant = GLib.Variant("a{sv}", { "MESSAGE": GLib.Variant("s", text) })
    GLib.log_variant(APPLICATION_ID, log_level, variant)

# searches a widget and its children for a specific buildable id or widget type
# note: stops at first match
def find_widget(parent, name="", widgetClass=None):
    p_name = Gtk.Buildable.get_name(parent)
    if p_name and name and p_name.strip() == name.strip():
        return parent

    if widgetClass and isinstance(parent, widgetClass):
        return parent

    if hasattr(parent, "get_children"):
        children = parent.get_children()
        for child in children:
            res = find_widget(child, name, widgetClass)
            if res is not None:
                return res

    return None

class Blueberry(Gtk.Application):
    def do_startup(self):
        Gtk.Application.do_startup(self)

        self.settings = Gio.Settings(schema="org.blueberry")

        #detect current DE
        wm_info = subprocess.getoutput("wmctrl -m")
        if "XDG_CURRENT_DESKTOP" in os.environ:
            xdg_current_desktop = os.environ["XDG_CURRENT_DESKTOP"]
        else:
            xdg_current_desktop = ""

        if "Marco" in wm_info or xdg_current_desktop == "MATE":
            self.de = "Mate"
            self.configuration_tools = {"sound": "mate-volume-control", "keyboard": "mate-keyboard-properties", "mouse": "mate-mouse-properties"}
        elif "Xfwm4" in wm_info or xdg_current_desktop == "XFCE":
            self.de = "Xfce"
            self.configuration_tools = {"keyboard": "xfce4-keyboard-settings", "mouse": "xfce4-mouse-settings"}
            if os.path.exists("/usr/bin/pavucontrol"):
                self.configuration_tools["sound"] = "pavucontrol"
            else:
                self.configuration_tools["sound"] = "xfce4-mixer"
        elif "Muffin" in wm_info or xdg_current_desktop == "X-Cinnamon":
            self.de = "Cinnamon"
            self.configuration_tools = {"sound": "cinnamon-settings sound", "keyboard": "cinnamon-settings keyboard", "mouse": "cinnamon-settings mouse"}
        elif "Mutter" in wm_info or "GNOME" in xdg_current_desktop:
            self.de = "Gnome"
            self.configuration_tools = {"sound": "gnome-control-center sound", "keyboard": "gnome-control-center keyboard", "mouse": "gnome-control-center mouse"}
        elif "Unity" in wm_info or xdg_current_desktop == "Unity":
            self.de = "Unity"
            self.configuration_tools = {"sound": "unity-control-center sound", "keyboard": "unity-control-center keyboard", "mouse": "unity-control-center mouse"}
        elif xdg_current_desktop == "LXDE":
            self.de = "LXDE"
            self.configuration_tools = {"sound": "pavucontrol", "keyboard": "lxinput", "mouse": "lxinput"}
        else:
            self.de = "Unknown"
            log("DE could not be detected!")
            self.configuration_tools = {}
            if os.path.exists("/usr/bin/pavucontrol"):
                self.configuration_tools["sound"] = "pavucontrol"

    def do_activate(self):
        if self.settings.get_boolean("tray-enabled"):
            subprocess.Popen(['blueberry-tray'])

        if len(self.get_windows()) > 0:
            # Blueberry is already running, focus the window
            self.get_active_window().present()
            return

        builder = Gtk.Builder.new_from_file(os.path.join(SCRIPT_DIR, "blueberry.ui"))

        window = builder.get_object("window")
        window.set_title(_("Bluetooth"))

        builder.get_object("settings-button").set_tooltip_text(_("Settings"))

        self.stack = builder.get_object("stack")

        debug = False
        if len(sys.argv) > 1 and sys.argv[1] == "debug":
            debug = True

        self.header_icon = builder.get_object("header-icon")
        self.status_icon = builder.get_object("status-icon")
        self.status_label = builder.get_object("status-label")

        # Devices
        self.rf_switch = builder.get_object("bluetooth-switch")
        self.rfkill = rfkillMagic.Interface(self.update_ui_callback, debug)
        self.rf_handler_id = self.rf_switch.connect("state-set", self.on_switch_changed)

        self.rfkill_error_image = builder.get_object("rfkill-error-image")

        # Settings
        settings_box = SettingsBox()
        settings_container = builder.get_object("settings-container")
        settings_container.pack_start(settings_box, True, True, 0)

        self.adapter_name_entry = Gtk.Entry(width_chars=30)
        self.adapter_name_entry.connect("focus-out-event", self.update_name_from_entry)
        self.adapter_name_entry.connect("activate", self.update_name_from_entry)
        self.adapter_name_entry.set_sensitive(False)
        self.adapter_name_entry.set_placeholder_text(_("Enable Bluetooth to edit"))
        row = SettingsRow(Gtk.Label(label=_("Name")), self.adapter_name_entry)
        row.set_tooltip_text(_("This is the Bluetooth name of your computer"))
        settings_box.add_row(row)

        self.obex_switch = Gtk.Switch()
        self.obex_switch.set_active(self.settings.get_boolean("obex-enabled"))
        self.obex_switch.connect("notify::active", self.on_obex_switch_toggled)
        self.settings.connect("changed", self.on_settings_changed)
        row = SettingsRow(Gtk.Label(label=_("Receive files from remote devices")), self.obex_switch)
        row.set_tooltip_text(_("This option allows your computer to receive files transferred over Bluetooth (OBEX)"))
        settings_box.add_row(row)

        self.tray_switch = Gtk.Switch()
        self.tray_switch.set_active(self.settings.get_boolean("tray-enabled"))
        self.tray_switch.connect("notify::active", self.on_tray_switch_toggled)
        self.settings.connect("changed", self.on_settings_changed)
        settings_box.add_row(SettingsRow(Gtk.Label(label=_("Show a tray icon")), self.tray_switch))

        settings_box.show_all()

        self.lib_widget = GnomeBluetooth.SettingsWidget.new()
        self.lib_widget.connect("panel-changed", self.panel_changed)
        builder.get_object("bluetooth-widget-box").pack_start(self.lib_widget, True, True, 0)

        self.add_window(window)
        window.show_all()
        self.update_ui_callback()

        # attempt to apply overrides and if we fail don't setup update hooks
        if self.get_label_widget_and_spinner(self.lib_widget):
            self.label_widget.set_text("")
            self.client = GnomeBluetooth.Client()
            self.model = self.client.get_model()
            self.model.connect('row-changed', self.on_adapter_status_changed)
            self.model.connect('row-deleted', self.on_adapter_status_changed)
            self.model.connect('row-inserted', self.on_adapter_status_changed)
            self.on_adapter_status_changed(self.lib_widget)

    def bluetooth_on(self):
        return not self.rfkill.soft_block and not self.rfkill.hard_block

    def get_label_widget_and_spinner(self, widget):
        # We attempt to override the widget style to replace
        # the explanation label text and destroy the spinner.
        # gnome_bluetooth_settings_widget doesn't give explicit access
        # to its label via gi so we recurse its child widgets to try
        # to find the parts we want to modify.

        # if the override fails for any reason it is disabled. update
        # signals in the main class are only connected if a test call
        # to this succeeds the first time

        self.label_widget = None
        self.spinner = None

        try:
            # not finding the label is fatal as it's our main purpose here
            self.label_widget = find_widget(widget, "explanation-label")

            if self.label_widget is None:
                raise LookupError("unable to find label to override")

            spinner = find_widget(widget, widgetClass=Gtk.Spinner)
            if spinner != None:
                spinner.destroy()

        except Exception as e:
            log("could not fetch label widget: {}".format(e))
            return False

        return True

    def on_adapter_status_changed(self, settings, foo=None, data=None):
        if self.bluetooth_on():
            self.adapter_name_entry.set_text(self.get_adapter_name())
        else:
            self.adapter_name_entry.set_text("")
        self.update_status(self.adapter_name_entry)

    def panel_changed(self, widget, panel=None):
        if not panel in self.configuration_tools:
            log("No configuration tool known for panel '{}'".format(panel))
        else:
            os.system("%s &" % self.configuration_tools[panel])

    def on_tray_switch_toggled(self, widget, data=None):
        if widget.get_active():
            self.settings.set_boolean("tray-enabled", True)
            subprocess.Popen(["blueberry-tray"])
        else:
            self.settings.set_boolean("tray-enabled", False)

    def on_obex_switch_toggled(self, widget, data=None):
        if widget.get_active():
            self.settings.set_boolean("obex-enabled", True)
            os.system("/usr/lib/blueberry/blueberry-obex-agent.py &")
        else:
            self.settings.set_boolean("obex-enabled", False)
            os.system("pkill -9 blueberry-obex");
        self.update_status()

    def update_name_from_entry(self, entry, arg1=None, data=None):
        name = entry.get_text()
        if name == "":
            adapter_name = self.client.props.default_adapter_name
            name = adapter_name if adapter_name else ""
            entry.set_text(name)

        subprocess.call(["bt-adapter", "--set", "Alias", name])
        self.update_status()

    def on_settings_changed(self, settings, key):
        self.tray_switch.set_active(self.settings.get_boolean("tray-enabled"))
        self.obex_switch.set_active(self.settings.get_boolean("obex-enabled"))

    def get_adapter_name(self):
        if not self.bluetooth_on():
            return ""

        name = None

        try:
            output = subprocess.check_output(["timeout", "2s", "bt-adapter", "-i"]).decode("utf-8").strip()
            for line in output.split("\n"):
                line = line.strip()
                if line.startswith("Alias: "):
                    name = line.replace("Alias: ", "").replace(" [rw]", "").replace(" [ro]", "")
                    break
        except Exception as cause:
            log("Could not retrieve the BT adapter name with 'bt-adapter -i': {}".format(cause))

        if name == None:
            default_name = self.client.props.default_adapter_name
            name = default_name if (default_name != None) else ""

        return name

    def update_status(self, path=None, iter=None, data=None):
        if not self.label_widget:
            return

        obex_enabled = self.settings.get_boolean("obex-enabled")

        if self.bluetooth_on():
            adapter_name = self.get_adapter_name()
            if adapter_name != "":
                self.adapter_name_entry.set_sensitive(True)

                if obex_enabled:
                    text = _("Visible as %s and available for Bluetooth file transfers.")
                else:
                    text = _("Visible as %s.")

                text = "%s\n" % text
                self.label_widget.set_markup(text % "\"%s\"" % adapter_name)
            else:
                self.adapter_name_entry.set_sensitive(False)
                self.label_widget.set_label("")
        else:
            self.adapter_name_entry.set_sensitive(False)

    def update_ui_callback(self):
        powered = False
        sensitive = False
        header_icon_name = status_icon_name = "blueberry-disabled"
        status_msg = ""

        if not self.rfkill.have_adapter:
            status_msg = _("No Bluetooth adapters found")
            status_icon_name = "dialog-info"
        elif self.rfkill.hard_block:
            status_msg = _("Bluetooth is disabled by hardware switch")
        elif self.rfkill.soft_block:
            status_msg = _("Bluetooth is disabled")
            sensitive = True
        else:
            header_icon_name = status_icon_name = "blueberry"
            sensitive = True
            powered = True

        # rfkill errors don't change powered state. If we forced the status page then the
        # most likely scenario would be a mixed state of powered-on bluetooth with UI
        # hiding that fact. Instead we show an error icon next to the switch and disable
        # the switch itself without changing any other state.
        if self.rfkill.rfkill_err is not None:
            strings = (_("An error has occurred"),
                       self.rfkill.rfkill_err.strip(),
                       _("Permission issues can be solved with a udev rule for rfkill or membership in the 'rfkill' group."))
            self.rfkill_error_image.set_tooltip_text("%s\n%s\n\n%s" % strings)
            self.rfkill_error_image.show()
            sensitive = False
        else:
            self.rfkill_error_image.hide()

        self.status_label.set_markup(status_msg)

        self.header_icon.set_from_icon_name(header_icon_name, Gtk.IconSize.LARGE_TOOLBAR)
        self.status_icon.set_from_icon_name(status_icon_name, Gtk.IconSize.DIALOG)

        self.rf_switch.set_sensitive(sensitive)

        self.rf_switch.handler_block(self.rf_handler_id)
        self.rf_switch.set_state(powered)
        self.rf_switch.handler_unblock(self.rf_handler_id)

        self.stack.set_visible_child_name("main-page" if powered else "status-page");

    def on_switch_changed(self, widget, state):
        self.rfkill.try_set_blocked(not state)
        return True

if __name__ == "__main__":
    app = Blueberry(application_id=APPLICATION_ID, flags=Gio.ApplicationFlags.FLAGS_NONE)
    app.run(None)
